docker/indexer/stages/processing/processing.go

```diff
@@ -123,7 +144,54 @@ func (s *Stage) processGit(ctx context.Context, repoInfo *preparation.Result) er
 		}
 		return nil
 	}); err != nil {
-		return err
+		return fmt.Errorf("failed during file walk: %v", err)
 	}
-	return s.Storer.Store(ctx, repoInfo, shared.MD5, fileResults)
+
+	log.Info("begin processing tree")
+	treeResults, bucketResults := processBuckets(fileResults)
+	log.Info("begin storage")
+	return s.Storer.Store(ctx, repoInfo, shared.MD5, bucketResults, treeResults)
+}
+
+// Returns bucket hashes and the individual file hashes of each bucket
+func processBuckets(fileResults []*FileResult) ([]*BucketNode, [][]*FileResult) {
+	buckets := make([][]*FileResult, bucketCount)
+
+	for _, fr := range fileResults {
+		// Evenly divide into bucketCount buckets,
+		idx := binary.BigEndian.Uint16(fr.Hash[0:2]) % bucketCount
+		buckets[idx] = append(buckets[idx], fr)
+	}
+
+	results := make([]*BucketNode, bucketCount)
+
+	for bucketIdx := range buckets {
+		// Sort hashes to produce deterministic bucket hashes
+		sort.Slice(buckets[bucketIdx], func(i, j int) bool {
+			for k := 0; k < len(buckets[bucketIdx][i].Hash); k++ {
+				if buckets[bucketIdx][i].Hash[k] < buckets[bucketIdx][j].Hash[k] {
+					return true
+				}
+				if buckets[bucketIdx][i].Hash[k] > buckets[bucketIdx][j].Hash[k] {
+					return false
+				}
+			}
+			return false
+		})
+
+		hasher := md5.New()
+		for _, v := range buckets[bucketIdx] {
+			_, err := hasher.Write(v.Hash)
+			if err != nil {
+				deflog.Panicf("Hasher error: %v", err)
```

Comment:
Can we instead just use log.Fatalf from glog?

Reply:
I don't think Fatalf calls deferred functions while panics do, we need that to acknowledge the messages.

Reply:
In cases like these, should we actually ack the message? It seems like we should retry in these cases.

Reply:
The current process is that all messages are acked because it won't show up in RepoIndex, causing the controller to queue it up to be ran again the next day. 

(This actually makes me wonder if there's a specific reason the ack is done at the end rather than immediately when pulling the message.)

But looking at the implementation of md5.Write(), err is actually not set, it's only needed to fulfill the io.Writer interface, so I can actually just remove the check.